x/spec/keeper/grpc_query_show_all_chains.go

```diff
@@ -10,20 +10,56 @@ import (
 )
 
 func (k Keeper) ShowAllChains(goCtx context.Context, req *types.QueryShowAllChainsRequest) (*types.QueryShowAllChainsResponse, error) {
+	var chainInfoList []*types.ShowAllChainsInfoStruct
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
 	if req == nil {
 		return nil, status.Error(codes.InvalidArgument, "invalid request")
 	}
 
-	ctx := sdk.UnwrapSDKContext(goCtx)
-	var names []string
-
 	// get all spec
 	allSpec := k.GetAllSpec(ctx)
 
-	// iterate over specs and extract the chain's name
+	// iterate over specs and extract the chains' info
 	for _, spec := range allSpec {
-		names = append(names, spec.GetName())
+		// get the spec's APIs
+		apis := spec.GetApis()
+
+		// create API interface names map to efficiently check if an interface already exists in it
+		apiInterfacesNamesMap := make(map[string]int)
+
+		// iterate over the APIs
+		for _, api := range apis {
```

Comment:
we are doing this code when staking a provider and verifying all of it's endpoints implement all apiInterfaces, can we reuse the functionality with inner functions rather repeat code?

Reply:
done